### PR TITLE
Wallet id salting

### PR DESCRIPTION
**Pull request by danimoh**
Hashing with a random salt that does not leave the hub to avoid that an
external app can derive wallet id's from public keys (Legacy and Ledger
accounts) or get a hint for private key guessing / brute forcing (for BIP39)
as hashing the private key is cheaper than deriving the public key.

### CODE REVIEW

File: src/lib/WalletStore.ts [hunk 1]
```diff
@@ -78,6 +59,34 @@ export class WalletStore {
         this._indexedDB = WalletStore.INDEXEDDB_IMPLEMENTATION;
     }
 
+    public async deriveId(keyId: string): Promise<string> {
```
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/WalletStore.ts]: I am in favour of it being non-static. All other exposed methods aren't so I think this is a good change.

File: src/lib/WalletStore.ts [hunk 2]
```diff
@@ -78,6 +59,34 @@ export class WalletStore {
...
+        const keyIdHash = Nimiq.Hash.computeBlake2b(saltedKeyIdBytes);
+
+        for (let i = 0; i <= (keyIdHash.length - WalletStore.WALLET_ID_LENGTH); i++) {
+            const id = Nimiq.BufferUtils.toHex(keyIdHash.subarray(i, i + WalletStore.WALLET_ID_LENGTH));
+            if (existingIds.indexOf(id) === -1) return id;
+        }
+
+        // Could not find an available wallet ID in the searched space.
+        // Hash again and recurse.
+        return this.deriveId(Nimiq.Hash.blake2b(keyIdHash).toBase64());
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/WalletStore.ts]: Indeed not necessary anymore.
I removed it.

File: src/lib/WalletStore.ts [hunk 3]
```diff
@@ -78,6 +59,34 @@ export class WalletStore {
...
+        const existingIds = wallets.map((wallet) => wallet.id);
+        const keyIdBytes = Nimiq.BufferUtils.fromBase64(keyId);
+
+        // Hashing with a random salt that does not leave the hub to avoid that an external app can derive wallet id's
+        // from public keys (Legacy and Ledger accounts) or get a hint for private key guessing / brute forcing (for
+        // BIP39) as hashing the private key is cheaper than deriving the public key.
+        const salt = await this._getSalt();
+        const saltedKeyIdBytes = new Uint8Array(keyIdBytes.length + salt.length);
+        saltedKeyIdBytes.set(keyIdBytes, 0);
+        saltedKeyIdBytes.set(salt, keyIdBytes.length);
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/WalletStore.ts]: It's just not necessary, can use the native methods.

File: src/lib/WalletStore.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/WalletStore.ts]: Yes, making it non-static feels hacky.

File: tests/unit/WalletStore.spec.ts [hunk 4]
```diff
@@ -35,10 +35,13 @@ const DUMMY: WalletInfo[] = [ // IDs must be alphabetically ordered, as the Wall
             1, 120, 3000e5,
         )], WalletType.LEDGER),
 ];
+const DUMMY_SALT = new Uint8Array(WalletStore.SALT_LENGTH);
```
[the remark this answers is not in the document]
Comment (danimoh) [reply tests/unit/WalletStore.spec.ts]: It doesn't really matter.
Could change it, but as I don't want to recompute the dummy wallet id's, I'm leaving it like this.

**Pull request merged**